File: test/swipe.test.ts
/**
 * Swiping: between tabs, and throwing a card off the screen.
 *
 *   npm run test:swipe
 *
 * All thresholds, and thresholds are where a gesture stops feeling like one.
 * The two failures that matter are opposites: too eager and a list can't be
 * scrolled without changing tabs, too strict and the swipe seems broken. Both
 * are decided here, so both are pinned here.
 */

import {
  BACK_DIRECTION,
  dismissedByDrag,
  OWNS_SWIPE,
  ownsItsSwipe,
  ROW_OPEN_AT,
  ROW_REVEAL,
  rowOffset,
  rowOpens,
  DISMISS_FLICK_PIXELS,
  DISMISS_PIXELS,
  DOMINANCE,
  FLICK_PIXELS,
  MIN_PIXELS,
  nextTab,
  startedAtEdge,
  swipeVerdict,
  TAB_ORDER,
  type Gesture,
} from '@/lib/swipe';

let failures = 0;

function check(label: string, ok: boolean, detail = '') {
  if (!ok) failures++;
  console.log(`${ok ? 'PASS' : 'FAIL'}  ${label}${detail ? `  — ${detail}` : ''}`);
}

const W = 390; // a common phone width

/*
  Enough of an element tree for ownsItsSwipe, which only ever asks two things:
  does this carry the attribute, and what is its parent. Building it by hand
  keeps this file free of a DOM implementation for one function.
*/
type Node = { hasAttribute: (n: string) => boolean; parentElement: Node | null };
const fakeRoot = (): Node => ({ hasAttribute: () => false, parentElement: null });
const fakeNode = (owns: boolean): Node => {
  const root = fakeRoot();
  const row = { hasAttribute: (n: string) => owns && n === OWNS_SWIPE, parentElement: root };
  return { hasAttribute: () => false, parentElement: row };
};
const g = (over: Partial<Gesture> = {}): Gesture => ({
  dx: 0,
  dy: 0,
  elapsed: 400,
  width: W,
  ...over,
});

function main() {
  /* ---------------------------------------------------------- the order */

  /*
    Every tab in the bar, in the bar's order. Subscriptions was added to the
    bar and not to this list, so swiping left from Items skipped straight to
    Settings — past a tab that was right there on screen.
  */
  check('four tabs, in bar order', TAB_ORDER.join() === 'home,items,subs,settings');
  check('swiping left goes forward', nextTab('home', 'left') === 'items');
  check('and again', nextTab('items', 'left') === 'subs');
  check('and again', nextTab('subs', 'left') === 'settings');
  check('swiping right goes back', nextTab('settings', 'right') === 'subs');

  // No wrapping. Settings → left → Home would answer "next" by jumping to the
  // far end of the app, and the bottom bar already shows there is no next.
  check('nothing past the last tab', nextTab('settings', 'left') === null);
  check('nothing before the first', nextTab('home', 'right') === null);

  /* -------------------------------------------------------- the verdict */

  // A deliberate drag: a fifth of the screen.
  check('a long drag left', swipeVerdict(g({ dx: -120 })) === 'left');
  check('a long drag right', swipeVerdict(g({ dx: 120 })) === 'right');
  check('half a screen still counts', swipeVerdict(g({ dx: -195 })) === 'left');

  // A flick: short and fast, and demanding full distance from it is what
  // makes a gesture feel heavy.
  check('a fast flick counts', swipeVerdict(g({ dx: -50, elapsed: 120 })) === 'left');
  check(
    'the same distance taken slowly does not',
    swipeVerdict(g({ dx: -50, elapsed: 900 })) === null,
  );

  /* ------------------------------------------------- what must not fire */

  // The common accident: a slightly diagonal thumb while reading a list.
  check('a vertical scroll is not a swipe', swipeVerdict(g({ dx: -30, dy: -200 })) === null);
  check(
    'nor a mostly-vertical diagonal',
    swipeVerdict(g({ dx: -90, dy: -140 })) === null,
    'dominance',
  );
  check(
    'a clearly horizontal diagonal does fire',
    swipeVerdict(g({ dx: -140, dy: -40 })) === 'left',
  );

  // Right at the dominance boundary, from both sides.
  const dy = 60;
  check('just under dominance is refused', swipeVerdict(g({ dx: -(dy * DOMINANCE - 1), dy })) === null);
  check(
    'just over it is allowed, given the distance',
    swipeVerdict(g({ dx: -140, dy: 140 / DOMINANCE - 1 })) === 'left',
  );

  check('a tap is not a swipe', swipeVerdict(g({ dx: 0, dy: 0, elapsed: 90 })) === null);
  check('a twitch is not a swipe', swipeVerdict(g({ dx: -12, elapsed: 100 })) === null);

  /* ------------------------------------------------- what a thumb does */

  // A real swipe is an arc, not a straight line — the thumb pivots. These are
  // the shapes that were being rejected while feeling, to the user, like
  // perfectly ordinary swipes.
  check('an arcing swipe counts', swipeVerdict(g({ dx: -110, dy: 62, elapsed: 260 })) === 'left');
  check(
    'a modest, unhurried swipe counts',
    swipeVerdict(g({ dx: -66, dy: 18, elapsed: 700 })) === 'left',
    'a fifth of the screen was too much to ask',
  );
  check(
    'a swipe that drifts up counts too',
    swipeVerdict(g({ dx: 95, dy: -50, elapsed: 300 })) === 'right',
  );

  /* ------------------------------------------------------ narrow screens */

  // The threshold scales, so a small phone doesn't demand a heroic drag and a
  // tablet doesn't fire on a nudge.
  check('a short screen needs less', swipeVerdict(g({ dx: -70, width: 320, elapsed: 800 })) === 'left');
  check(
    'a wide screen needs more',
    swipeVerdict(g({ dx: -70, width: 1024, elapsed: 800 })) === null,
  );

  /* ----------------------------------------------------------- the edges */

  // Android's back gesture owns the edges. One gesture must not do two things.
  check('the left edge belongs to the system', startedAtEdge(8, W));
  check('so does the right', startedAtEdge(W - 8, W));
  check('the middle is ours', !startedAtEdge(W / 2, W));
  check('and just inside the guard', !startedAtEdge(40, W));

  /* ------------------------------------------------------- stepping back */

  /*
    A pushed screen — an item, the rooms list, the bin — reads a horizontal
    swipe as "up one level". Getting the sign backwards would send people
    deeper, or nowhere, so the direction is named in one place and checked
    against the verdict that produces it.
  */
  check('back is to the right', BACK_DIRECTION === 'right');
  check(
    'and a rightward drag produces it',
    swipeVerdict(g({ dx: 120 })) === BACK_DIRECTION,
  );
  check('dragging the other way is not back', swipeVerdict(g({ dx: -120 })) !== BACK_DIRECTION);

  // Our gesture and Android's system back must not both fire. The edge belongs
  // to the system, and useSwipeNav refuses to start there — which is what
  // makes it safe to put a back swipe on the same screens.
  check('a swipe from the edge is never ours', startedAtEdge(10, W) && startedAtEdge(W - 10, W));

  /* ------------------------------------------------ pushing a row aside */

  /*
    Left only, and never mistakable for a scroll. The row slides to reveal a
    delete button — it does not delete — because a horizontal drag is the
    easiest gesture in the app to make by accident while thumbing down a list.
  */
  check('a decisive left drag opens it', rowOpens(-60, 4));
  check('a small one does not', !rowOpens(-12, 0));
  check('rightwards never opens it', !rowOpens(60, 0));
  check('a diagonal scroll does not', !rowOpens(-40, 90));
  check('right at the threshold', !rowOpens(-ROW_OPEN_AT, 0) && rowOpens(-ROW_OPEN_AT - 1, 0));

  // Opening is cheaper to undo than changing tabs, so it asks for less.
  check('it takes less than a tab swipe', ROW_OPEN_AT < MIN_PIXELS, `${ROW_OPEN_AT} vs ${MIN_PIXELS}`);

  /*
    The row and the shell were both reading the same drag: the row slid open
    and the app changed tab underneath it. A row marked as owning its swipe is
    skipped by the tab gesture entirely.
  */
  check(
    'a row that owns its swipe is left alone',
    ownsItsSwipe(fakeNode(true), fakeRoot()) && !ownsItsSwipe(fakeNode(false), fakeRoot()),
  );

  check('the row never travels past the button', rowOffset(-400, false) === -ROW_REVEAL);
  check('nor right of where it started', rowOffset(200, false) === 0);
  check('mid-drag it tracks the finger', rowOffset(-40, false) === -40);
  check('an open row drags from where it is', rowOffset(0, true) === -ROW_REVEAL);
  check('and can be pushed back shut', rowOffset(ROW_REVEAL, true) === 0);

  /* ------------------------------------------------ throwing a card away */

  /*
    The paper reminder after every save. Deliberately looser than the tab
    swipe, and the asymmetry is the argument: changing tabs by accident loses
    your place, closing a reminder by accident costs nothing at all.
  */
  const d = (dy: number, elapsed = 500) => dismissedByDrag({ dy, elapsed });

  check('a firm drag down closes it', d(DISMISS_PIXELS));
  check('and so does one up — the card is centred, not anchored', d(-DISMISS_PIXELS));
  check('a nudge does not', !d(12));
  check('nor does a slow short drag', !d(DISMISS_FLICK_PIXELS + 2, 900));
  check('but a flick that short does', d(DISMISS_FLICK_PIXELS + 2, 200));
  check('a flick upward too', d(-(DISMISS_FLICK_PIXELS + 2), 200));
  check('a tap is not a throw', !d(0, 90));

  // Looser than the tab gesture, checked rather than asserted in a comment:
  // the day someone tunes one of these, the relationship should hold.
  check(
    'it takes less than changing tabs',
    DISMISS_PIXELS < MIN_PIXELS && DISMISS_FLICK_PIXELS < FLICK_PIXELS,
    `${DISMISS_PIXELS} vs ${MIN_PIXELS}, ${DISMISS_FLICK_PIXELS} vs ${FLICK_PIXELS}`,
  );

  console.log(failures === 0 ? '\nall green' : `\n${failures} failure(s)`);
  process.exit(failures === 0 ? 0 : 1);
}

main();

File: mobile/test/swipe_test.dart
/// Swipe thresholds.
///
///   dart test test/swipe_test.dart
///
/// Translated from `test/swipe.test.ts`. Every number here is a judgement that
/// shows up as a feel, which is exactly why they are asserted rather than
/// buried in an event handler.
library;

import 'package:stash_it/logic/swipe.dart';
import 'package:flutter_test/flutter_test.dart';

const phone = 390.0; // a normal handset width

Gesture swipe(double dx, {double dy = 0, int ms = 200, double width = phone}) =>
    Gesture(dx: dx, dy: dy, elapsed: Duration(milliseconds: ms), width: width);

void main() {
  group('which tab', () {
    /*
      Subscriptions was once added to the bottom bar and not to the order list,
      so swiping left from Items landed on Settings — it skipped the tab
      sitting between them and nothing on screen explained why. An enum makes
      that a compile error; this asserts the order anyway.
    */
    test('the order matches the bar', () {
      expect(Tab.values,
          [Tab.home, Tab.items, Tab.subs, Tab.papers, Tab.settings]);
    });

    /*
      The three tests that were here walked `nextTab`, which the shell no
      longer calls: the pages are dragged directly and the `PageView` walks
      this list itself, ends included.

      So the order above is the whole contract now — it is what the bar draws,
      what the swipe traverses, and what a tap animates to.
    */
  });

  group('was that a swipe', () {
    test('a long deliberate drag is', () {
      // A fifth of the screen, slowly.
      expect(swipeVerdict(swipe(-120, ms: 900)), Direction.left);
    });

    test('a short fast flick is too', () {
      // Nowhere near a fifth of the screen, but unmistakably deliberate.
      expect(swipeVerdict(swipe(-40, ms: 120)), Direction.left);
    });

    test('but a short slow drag is not', () {
      // Neither far enough to be deliberate nor fast enough to be a flick.
      expect(swipeVerdict(swipe(-40, ms: 900)), isNull);
    });

    test('a tiny movement is never one', () {
      expect(swipeVerdict(swipe(-20, ms: 60)), isNull);
    });

    test('direction follows the finger', () {
      // Content follows the finger, so dragging left reveals what is right.
      expect(swipeVerdict(swipe(-120, ms: 900)), Direction.left);
      expect(swipeVerdict(swipe(120, ms: 900)), Direction.right);
    });

    /*
      THE ACCIDENT WORTH PREVENTING. Skimming a long list with the thumb is a
      diagonal drag, and at 1.3 dominance it still does not change tabs.
    */
    test('a diagonal scroll is left alone', () {
      expect(swipeVerdict(swipe(-100, dy: 90, ms: 900)), isNull);
    });

    test('but a real swipe with a thumb arc is not', () {
      expect(swipeVerdict(swipe(-140, dy: 40, ms: 900)), Direction.left);
    });

    test('the threshold scales with the screen', () {
      // 90px is over a fifth of a small phone and under a fifth of a tablet.
      expect(swipeVerdict(swipe(-90, ms: 2000, width: 360)), Direction.left);
      expect(swipeVerdict(swipe(-90, ms: 2000, width: 1000)), isNull);
    });

    test('and never drops below the pixel floor', () {
      // A very narrow window must still demand a real movement.
      expect(swipeVerdict(swipe(-40, ms: 2000, width: 100)), isNull);
    });
  });

  group('a row sliding aside', () {
    test('a firm leftward drag opens it', () {
      expect(rowOpens(-40, 0), isTrue);
    });

    // Strictly past the threshold, not at it — a boundary that reads one way
    // in prose and another in code is one somebody will get wrong later.
    test('exactly at the threshold does not', () {
      expect(rowOpens(-rowOpenAt, 0), isFalse);
    });

    test('rightwards never does', () {
      expect(rowOpens(40, 0), isFalse);
      expect(rowOpens(200, 0), isFalse);
    });

    test('nor does a mostly-vertical drag', () {
      expect(rowOpens(-40, 60), isFalse);
    });

    test('the row never slides right of home', () {
      expect(rowOffset(50, false), 0);
    });

    test('nor past the button', () {
      expect(rowOffset(-500, false), -rowReveal);
      expect(rowOffset(-500, true), -rowReveal);
    });

    test('and follows the finger in between', () {
      expect(rowOffset(-40, false), -40);
      // Already open, dragging back towards home.
      expect(rowOffset(40, true), -rowReveal + 40);
    });

    test('opening is easier than the reveal is wide', () {
      // Opening costs nothing, so a hesitant drag should succeed.
      expect(rowOpenAt, lessThan(rowReveal / 2));
    });
  });

  group('throwing a card away', () {
    test('a firm drag either way dismisses', () {
      expect(dismissedByDrag(Drag(dy: 50, elapsed: const Duration(seconds: 2))),
          isTrue);
      expect(
          dismissedByDrag(Drag(dy: -50, elapsed: const Duration(seconds: 2))),
          isTrue);
    });

    test('a quick flick does too', () {
      expect(
          dismissedByDrag(
              const Drag(dy: 30, elapsed: Duration(milliseconds: 150))),
          isTrue);
    });

    test('a small slow nudge does not', () {
      expect(dismissedByDrag(const Drag(dy: 30, elapsed: Duration(seconds: 2))),
          isFalse);
    });

    /*
      THE ASYMMETRY, ASSERTED. Changing tabs by accident loses your place;
      closing a reminder by accident costs nothing, and the reminder appears
      after every save. So dismissing must be looser — and looser than the
      *floor*, not merely looser than the tab threshold on whatever handset
      this was tuned against, since that one scales with the screen.
    */
    test('and it is looser than the tab swipe on every device', () {
      expect(dismissPixels, lessThan(minPixels));
      expect(dismissFlickPixels, lessThan(flickPixels));
    });
  });

  group('the screen edges', () {
    // The system back gesture owns them on Android. Ours must not also fire,
    // or one gesture does two things.
    test('the left edge belongs to the system', () {
      expect(startedAtEdge(10, phone), isTrue);
    });

    test('so does the right', () {
      expect(startedAtEdge(phone - 10, phone), isTrue);
    });

    test('the middle is ours', () {
      expect(startedAtEdge(phone / 2, phone), isFalse);
    });

    test('and the guard is narrow enough to leave a screen behind', () {
      expect(edgeGuard * 2, lessThan(phone / 3));
    });
  });
}
